### PR TITLE
OTT-1064: Support for lineitems with dealid targeting

### PR DESCRIPTION
**Pull request by Pubmatic-Dhruv-Sonone**
Deal lineitems with deal id targeting
1) Added new setting  DEAL_CONFIG_TYPE  - "DEALID" or "DEALTIER"
2) Changes to create sponsorship lineitem with deal id targeting
3) Summary fix
4) No of line items = number of deals
5) Lineitems for all slot will be under one order.
6) Updated the setup doc.

### CODE REVIEW

File: docs/adpod_setup.md [hunk 1]
```diff
@@ -83,26 +84,35 @@ Ex: https://ow.pubmatic.com/cache?uuid=123456789.
...
 1. Single order will have lineitems for all the slots.
-2. Line items will be created with `pwtdt` targeting. Optional `pwtpid` (bidder targeting) is added if bidder codes are set.
-3. Dealtier config should be similar to what is expected in openwrap bid request and response.
+2. Line items will be created with `pwtdt` targeting for dealtier lineitems and `pwtdid` targeting for dealid lineitems . Optional `pwtpid` (bidder targeting) is added if bidder codes are set.
+3. Deal config should be similar to what is expected in openwrap bid request and response.
 4. For dealtier lineitems:
    <br> No of lineitems created = len(dealpriority)*len(prefix) * no of bidders
    <br>Ex: For config `{"appnexus":{"price":20,"prefix":["appx"],"dealpriority":[8,10]},"pubmatic":{"price":10,"prefix":["pubm"],"dealpriority":[5]}}`
    <br><br>No of lineitems for pubmatic:  `1` --> `["pubm5"]` -->  `price:10`
    <br>No of lineitems for appnexus:  `2`  --> `["pubm8","pubm10"]` --> `price:20`
```
Comment (pm-isha-bharti) [inline docs/adpod_setup.md:109]: Should the prefix for `appnexus` be `appx`?

File: docs/adpod_setup.md [hunk 2]
```diff
@@ -83,26 +84,35 @@ Ex: https://ow.pubmatic.com/cache?uuid=123456789.
...
    <br><br> Targeting keys(`pwtdt`): `["pubm5", "pubm8", "pubm10"]` 
-                
+5. For dealid lineitems:
+   <br> No of lineitems created = len(dealids) * no of bidders
+   <br>Ex: For config `{"appnexus":{"price":20,"dealids":["ApnxDeal1"]},"pubmatic":{"price":10,"dealids":["PubmDeal1"]}}`
+   <br><br>No of lineitems for pubmatic:  `1` --> `["PubmDeal1"]` -->  `price:10`
+   <br>No of lineitems for appnexus:  `2`  --> `["ApnxDeal1"]` --> `price:20`
```
Comment (pm-isha-bharti) [inline docs/adpod_setup.md:115]: The number of LI for Appnexus will be 1 right? Or `2` is representing index here?

File: settings.py [hunk 3]
```diff
@@ -189,16 +189,32 @@
 # THis option is only for Adpod setup
 # Set ENABLE_DEAL_LINEITEM = True for creating deal line item
 # Defaults to False
-ENABLE_DEAL_LINEITEM = False
+ENABLE_DEAL_LINEITEM = True
```
Comment (pm-isha-bharti) [inline settings.py:192]: Should the default value be set to `False` as publisher might not want to enable deal setup always.

File: tasks/add_new_openwrap_partner.py [hunk 4]
```diff
@@ -1185,7 +1240,10 @@ def main():
 
   if  (setup_type == constant.ADPOD and deal_lineitem_enabled == True) and lineitem_type != constant.LI_SPONSORSHIP:
      raise  BadSettingException('DFP_LINEITEM_TYPE should be SPONSORSHIP for creating deal lineitems')
- 
+
+  if (setup_type == constant.ADPOD and deal_lineitem_enabled == True) and (deal_config_type is None or (deal_config_type != constant.DEALTIER and deal_config_type != constant.DEALID)):
+     raise  MissingSettingException('DEAL_CONFIG_TYPE should be DEALTIER or DEALID')
```
Comment (pm-isha-bharti) [inline tasks/add_new_openwrap_partner.py:1245]: We can use the constants defined in the error: 
`constant.DEALTIER `and `constant.DEALID`

File: docs/adpod_setup.md (hunk 1 above)
Comment (Pubmatic-Dhruv-Sonone) [reply docs/adpod_setup.md]: Fixed.

File: docs/adpod_setup.md (hunk 2 above)
Comment (Pubmatic-Dhruv-Sonone) [reply docs/adpod_setup.md]: Changed to **1**

File: docs/adpod_setup.md [hunk 5]
```diff
@@ -83,26 +84,35 @@ Ex: https://ow.pubmatic.com/cache?uuid=123456789.
...
-2. Line items will be created with `pwtdt` targeting. Optional `pwtpid` (bidder targeting) is added if bidder codes are set.
-3. Dealtier config should be similar to what is expected in openwrap bid request and response.
+2. Line items will be created with `pwtdt` targeting for dealtier lineitems and `pwtdid` targeting for dealid lineitems . Optional `pwtpid` (bidder targeting) is added if bidder codes are set.
+3. Deal config should be similar to what is expected in openwrap bid request and response.
 4. For dealtier lineitems:
    <br> No of lineitems created = len(dealpriority)*len(prefix) * no of bidders
    <br>Ex: For config `{"appnexus":{"price":20,"prefix":["appx"],"dealpriority":[8,10]},"pubmatic":{"price":10,"prefix":["pubm"],"dealpriority":[5]}}`
    <br><br>No of lineitems for pubmatic:  `1` --> `["pubm5"]` -->  `price:10`
    <br>No of lineitems for appnexus:  `2`  --> `["pubm8","pubm10"]` --> `price:20`
    <br><br> Targeting keys(`pwtdt`): `["pubm5", "pubm8", "pubm10"]`
```
[the remark this answers is not in the document]
Comment (Pubmatic-Dhruv-Sonone) [reply docs/adpod_setup.md]: Done

File: settings.py (hunk 3 above)
Comment (Pubmatic-Dhruv-Sonone) [reply settings.py]: Changed

File: tasks/add_new_openwrap_partner.py (hunk 4 above)
Comment (Pubmatic-Dhruv-Sonone) [reply tasks/add_new_openwrap_partner.py]: Done

**Review by pm-isha-bharti — APPROVED** [review]

**Pull request merged**